graphql_compiler/schema_merging/rename_schema.py

```diff
@@ -0,0 +1,232 @@
+# Copyright 2019-present Kensho Technologies, LLC.
+from collections import namedtuple
+
+from graphql import build_ast_schema, parse
+from graphql.language.visitor import Visitor, visit
+
+from .utils import SchemaError, get_schema_data
+
+
+RenamedSchema = namedtuple(
+    'RenamedSchema', [
+        'schema_ast',  # type: Document, ast representing the renamed schema
+        'reverse_name_map',  # type: Dict[str, str], new type name to old type name
+        'reverse_root_field_map'  # type: Dict[str, str], new root field name to old root field
+    ]
+)
+
+
+def rename_schema(schema_string, rename_func=lambda name: name):
+    """Create a RenamedSchema, where types and query root fields are renamed using rename_func.
+
+    Args:
+        schema_string: string describing a valid schema that does not contain extensions
+        rename_func: callable that takes string to string, used to transform the names of
+                     types, interfaces, enums, and root fields. Defaults to identity
+
+    Returns:
+        RenamedSchema
+
+    Raises:
+        SchemaError if input schema_string does not represent a valid schema without extensions,
+            or if there are conflicts between the renamed types or root fields
+    """
+    # Check that the input string is a parseable and valid schema.
+    try:
+        ast = parse(schema_string)
```

Comment:
Thoughts on accepting an AST directly (i.e. making the caller parse the string) instead here? Curious about what we might stand to gain and lose by switching.

Reply:
If the function takes in an AST directly, I'm still going to make a copy of it before doing anything else since in the next steps I modify the AST. I like the idea of passing in a string also because the string is immutable.

Reply:
Related question, should the output RenamedSchema contain the schema AST, or the schema string? It's easy to go back and forth between them with `parse` and `print_ast`, and there are benefits and downsides either way. Dealing with strings is nice since strings are immutable. Dealing with ASTs is nice since manipulations happen on ASTs, and not converting to strings in between means there's less waste converting back and forth.